toFlow() and .watch() result in IllegalArgumentException: Flow has more than one element

### Version

3.7.3

### Summary

When using cache and network, both:

- `apollo.query(makeQuery()).fetchPolicy(FetchPolicy.CacheAndNetwork).toFlow()`
- `apollo.query(makeQuery()).fetchPolicy(FetchPolicy.CacheAndNetwork).watch()`

result in `java.lang.IllegalArgumentException: Flow has more than one element`.

If I use `FetchPolicy.NetworkOnly` no exception is thrown and the data is collected as expected. Am I using the `FetchPolicy` feature correctly? I assumed the flow would emit twice, once for the cache and one more time for the network. Likewise I assumed that `.watch()` will continuously emit whenever the cache updates, such as after a mutation in another part of the app that updates any of the data related to the original query.

I'm also noticing that after deleting the app and rerunning it, the sqlite database in the Device File Explorer is populated with data from a very old version of the schema, when the schema and data types were completely different. Could this be related? EDIT: For this cleaning and rebuilding the project fixed it. The flow exception still remains.

Ok! Thanks for your support. Here's what the stack trace printed:

```
HttpClient(_cause=java.lang.IllegalArgumentException: Flow has more than one element)
	at com.davidarvelo.app.core.interactors.BaseQueryInteractor$mapped$1.invokeSuspend(BaseQueryInteractor.kt:49)
	at com.davidarvelo.app.core.interactors.BaseQueryInteractor$mapped$1.invoke(Unknown Source:9)
	at com.davidarvelo.app.core.interactors.BaseQueryInteractor$mapped$1.invoke(Unknown Source:6)
	at kotlinx.coroutines.flow.FlowKt__ErrorsKt$catch$$inlined$unsafeFlow$1.collect(SafeCollector.common.kt:114)
	at kotlinx.coroutines.flow.FlowKt__ErrorsKt$catch$$inlined$unsafeFlow$1$1.invokeSuspend(Unknown Source:15)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7898)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: java.lang.IllegalArgumentException: Flow has more than one element
	at kotlinx.coroutines.flow.FlowKt__ReduceKt$single$2.emit(Reduce.kt:58)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:87)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:66)
	at com.apollographql.apollo3.cache.normalized.FetchPolicyInterceptors$CacheAndNetworkInterceptor$1$intercept$1$invokeSuspend$$inlined$map$1$2.emit(Emitters.kt:223)
	at kotlinx.coroutines.flow.FlowKt__ErrorsKt$catchImpl$2.emit(Errors.kt:158)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAllImpl$FlowKt__ChannelsKt(Channels.kt:62)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.access$emitAllImpl$FlowKt__ChannelsKt(Channels.kt:1)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAllImpl$1.invokeSuspend(Unknown Source:14)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	... 9 more
```

This is the contents of my `ApolloClient.Builder`:

```kotlin
private var appSettingsStorage: SettingsStorage? = null

object DataClient {

    private const val AUTHORIZATION_HEADER = "Authorization"

    private val memoryFirstThenSqlCacheFactory = MemoryCacheFactory(10 * 1024 * 1024)
    // I used to use this but it's buggy for now.
    // Seems SQLite is returning duplicate records and
    // Apollo cancels the flow because it expects a single record.
    // TODO: Maybe in a later version of Apollo (as of this writing, 3.7.3) I can re-enable this.
//    .chain(SqlNormalizedCacheFactory("habbit.db"))

    val apollo = ApolloClient.Builder()
        .serverUrl("http://localhost:4000/graphql")
        .addHttpInterceptor(AuthorizationInterceptor())
        // Note that mutations will always be sent as POST requests, regardless of these settings, as to avoid hitting caches.
        .autoPersistedQueries(
            // For the initial hashed query that does not send the actual Graphql document
            httpMethodForHashedQueries = HttpMethod.Get,
            // For the follow-up query that sends the full document if the initial hashed query was not found
            httpMethodForDocumentQueries = HttpMethod.Get
        )
        .normalizedCache(normalizedCacheFactory = memoryFirstThenSqlCacheFactory)
        .build()

    fun configure(settingsStorage: SettingsStorage) {
        appSettingsStorage = settingsStorage
    }

    class AuthorizationInterceptor : HttpInterceptor {
        override suspend fun intercept(request: HttpRequest, chain: HttpInterceptorChain): HttpResponse {
            val requestBuilder = request.newBuilder().let {
                val authToken = appSettingsStorage?.authToken
                if (authToken != null) {
                    it.addHeader(AUTHORIZATION_HEADER, "Bearer $authToken")
                } else {
                    it
                }
            }
            return chain.proceed(requestBuilder.build())
        }
    }
}
```

--- a/libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/ApolloClient.java
+++ b/libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/ApolloClient.java
@@ -171,6 +171,7 @@ public class ApolloClient implements Closeable {
     private WebSocket.Factory webSocketFactory;
     private Executor executor;
     private List<ApolloInterceptor> interceptors = new ArrayList<>();
+    private ApolloInterceptor apqInterceptor;
     private List<HttpInterceptor> httpInterceptors = new ArrayList<>();
     private WsProtocol.Factory wsProtocolFactory;
     private List<HttpHeader> wsHeaders = new ArrayList<>();
@@ -353,6 +354,38 @@ public class ApolloClient implements Closeable {
       return this;
     }
 
+
+    public Builder autoPersistedQueries() {
+      return autoPersistedQueries(HttpMethod.Get, HttpMethod.Post, true);
+    }
+
+    public Builder autoPersistedQueries(
+        HttpMethod httpMethodForHashedQueries
+    ) {
+      return autoPersistedQueries(httpMethodForHashedQueries, HttpMethod.Post, true);
+    }
+
+    public Builder autoPersistedQueries(
+        HttpMethod httpMethodForHashedQueries,
+        HttpMethod httpMethodForDocumentQueries
+    ) {
+      return autoPersistedQueries(httpMethodForHashedQueries, httpMethodForDocumentQueries, true);
+    }
+
+    public Builder autoPersistedQueries(
+        HttpMethod httpMethodForHashedQueries,
+        HttpMethod httpMethodForDocumentQueries,
+        boolean enableByDefault
+    ) {
+      apqInterceptor = new AutoPersistedQueryInterceptor(
+              httpMethodForHashedQueries,
+              httpMethodForDocumentQueries);
+
+      enableAutoPersistedQueries(enableByDefault);
+
+      return this;
+    }
+
     public ApolloClient build() {
       if (executor == null) {
         executor = defaultExecutor();
@@ -430,6 +463,9 @@ public class ApolloClient implements Closeable {
         }
       }
 
+      if (apqInterceptor != null) {
+        interceptors.add(apqInterceptor);
+      }
       return new ApolloClient(
           executor,
           networkTransport,
@@ -445,38 +481,6 @@ public class ApolloClient implements Closeable {
       );
     }
 
-    public Builder autoPersistedQueries() {
-      return autoPersistedQueries(HttpMethod.Get, HttpMethod.Post, true);
-    }
-
-    public Builder autoPersistedQueries(
-        HttpMethod httpMethodForHashedQueries
-    ) {
-      return autoPersistedQueries(httpMethodForHashedQueries, HttpMethod.Post, true);
-    }
-
-    public Builder autoPersistedQueries(
-        HttpMethod httpMethodForHashedQueries,
-        HttpMethod httpMethodForDocumentQueries
-    ) {
-      return autoPersistedQueries(httpMethodForHashedQueries, httpMethodForDocumentQueries, true);
-    }
-
-    public Builder autoPersistedQueries(
-        HttpMethod httpMethodForHashedQueries,
-        HttpMethod httpMethodForDocumentQueries,
-        boolean enableByDefault
-    ) {
-      addInterceptor(
-          new AutoPersistedQueryInterceptor(
-              httpMethodForHashedQueries,
-              httpMethodForDocumentQueries
-          )
-      );
-      enableAutoPersistedQueries(enableByDefault);
-
-      return this;
-    }
 
     /**
      * Batch HTTP queries to execute multiple at once. This reduces the number of HTTP round trips at the price of increased latency as

--- a/libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt
+++ b/libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt
@@ -186,6 +186,12 @@ private constructor(
     private var subscriptionNetworkTransport: NetworkTransport? = null
     private val customScalarAdaptersBuilder = CustomScalarAdapters.Builder()
     private val _interceptors: MutableList<ApolloInterceptor> = mutableListOf()
+
+    /**
+     * The APQ interceptor is special because it needs to be always last.
+     * If not, there is the risk that it retries cache queries which make no sense.
+     */
+    private var apqInterceptor: ApolloInterceptor? = null
     val interceptors: List<ApolloInterceptor> = _interceptors
     private val httpInterceptors: MutableList<HttpInterceptor> = mutableListOf()
     private var dispatcher: CoroutineDispatcher? = null
@@ -445,12 +451,11 @@ private constructor(
         httpMethodForDocumentQueries: HttpMethod = HttpMethod.Post,
         enableByDefault: Boolean = true,
     ) = apply {
-      addInterceptor(
-          AutoPersistedQueryInterceptor(
-              httpMethodForHashedQueries,
-              httpMethodForDocumentQueries
-          )
+      apqInterceptor = AutoPersistedQueryInterceptor(
+          httpMethodForHashedQueries,
+          httpMethodForDocumentQueries
       )
+
       enableAutoPersistedQueries(enableByDefault)
     }
 
@@ -576,7 +581,7 @@ private constructor(
           networkTransport = networkTransport,
           subscriptionNetworkTransport = subscriptionNetworkTransport,
           customScalarAdapters = customScalarAdaptersBuilder.build(),
-          interceptors = _interceptors,
+          interceptors = _interceptors + listOfNotNull(apqInterceptor),
           dispatcher = dispatcher,
           executionContext = executionContext,
           httpMethod = httpMethod,

--- a/tests/integration-tests/src/jvmTest/kotlin/test/ApqCacheTest.kt
+++ b/tests/integration-tests/src/jvmTest/kotlin/test/ApqCacheTest.kt
@@ -1,0 +1,49 @@
+package test
+
+import com.apollographql.apollo3.ApolloClient
+import com.apollographql.apollo3.api.http.HttpMethod
+import com.apollographql.apollo3.cache.normalized.FetchPolicy
+import com.apollographql.apollo3.cache.normalized.api.MemoryCacheFactory
+import com.apollographql.apollo3.cache.normalized.fetchPolicy
+import com.apollographql.apollo3.cache.normalized.normalizedCache
+import com.apollographql.apollo3.integration.normalizer.HeroNameQuery
+import com.apollographql.apollo3.mockserver.MockServer
+import com.apollographql.apollo3.testing.enqueue
+import com.apollographql.apollo3.testing.internal.runTest
+import org.junit.Test
+
+class ApqCacheTest {
+  /**
+   * https://github.com/apollographql/apollo-kotlin/issues/4617
+   */
+  @Test
+  fun apqAndCache() = runTest {
+    val mockServer = MockServer()
+
+    val data = HeroNameQuery.Data(HeroNameQuery.Hero("R2-D2"))
+    val query = HeroNameQuery()
+
+    mockServer.enqueue(query, data)
+    mockServer.enqueue(query, data)
+
+    val apolloClient = ApolloClient.Builder()
+        .serverUrl(mockServer.url())
+        // Note that mutations will always be sent as POST requests, regardless of these settings, as to avoid hitting caches.
+        .autoPersistedQueries(
+            // For the initial hashed query that does not send the actual Graphql document
+            httpMethodForHashedQueries = HttpMethod.Get,
+            // For the follow-up query that sends the full document if the initial hashed query was not found
+            httpMethodForDocumentQueries = HttpMethod.Get
+        )
+        .normalizedCache(normalizedCacheFactory = MemoryCacheFactory(10 * 1024 * 1024))
+        .build()
+
+    // put in the cache
+    apolloClient.query(HeroNameQuery()).fetchPolicy(FetchPolicy.NetworkOnly).execute()
+
+    // Query cache and network, it shouldn't throw
+    apolloClient.query(HeroNameQuery()).fetchPolicy(FetchPolicy.CacheAndNetwork).toFlow().collect {
+      println(it.data)
+    }
+  }
+}

